slippi/parse.py

```
@@ -116,7 +123,7 @@ def _parse_events(stream, payload_sizes, total_size, handlers):
     event = None
 
     # `total_size` will be zero for in-progress replays
-    while (total_size == 0 or bytes_read < total_size) and event != ParseEvent.END:
+    while ((total_size == 0 and stream.tell() < stream.getbuffer().nbytes) or bytes_read < total_size) and event != ParseEvent.END:
```

Before a single fragment would be parsed and stream.tell() < stream.getbuffer().nbytes was 76 < 76. Instead of cancelling, total_size == 0 kept it running whilst running into EofError.

---

This will break with non-seekable streams. I'm not clear on the situation you're describing; could you explain more?

---

Sure thing. I'm using this in an outside application and pass in Start, Pre, Post, and End frame updates but not together. The problem persists that I'll access _parse_events directly despite being "protected" with total_size == 0, this then tries to process the payload twice. First time is successful then the second causes EofError as the previous condition was (total_size == 0 or bytes_read < total_size) where total_size == 0 would cause it to run continuously.
I will say that I could say total_size should be 76 but I wanted it a bit more fitting.

Regarding non-seekable streams, what case would this arise? Doesn't BytesIO provide a seek?

---

Things like pipes or network sockets will be non-seekable. We don't necessarily have a `BytesIO` buffer until we read the data for a specific event in `_parse_event`, so we can't assume seekability outside that scope.

The `bytes_read < total_size` thing is there as an additional safeguard to prevent reading past where we're supposed to. The main termination condition is seeing an `END` event: `_parse_events` is intended to read from a `raw` block, which should always end with one of those.

I'm not sure I completely understand the use case you're describing, but if you're sending streams of events that don't look like a normal `raw` block, you have a few options:
1. Add an `END` event when you want to stop parsing, even if you're just synthesizing a bogus one.
1. Pass in an accurate `total_size`. This requires that you know the size in advance, of course.
1. Catch the `EofError`. This should work but might mask real errors, depending on your use case.
1. Don't use `_parse_events`, and instead call `_parse_event` in your own loop.

---

(2) seems like the better option, I'll cleanup and remote the extra guard update and payload changes.